weblate/utils/ratelimit.py

```diff
@@ -159,3 +159,57 @@ def _rate_wrap(
         return wraps(function)(_rate_wrap)
 
     return _session_ratelimit_post_controller
+
+
+class RateLimitCache:
+    """Cache wrapper for rate limiting with atomic check/decrement operations."""
+    
+    def __init__(self, base_key: str, attempts: int, window: int):
+        self.cache_key = f"notify:rate:{base_key}:{attempts}:{window}"
+        self.attempts = attempts
+        self.window = window
+        
+        # Initialize the bucket (atomically on redis) if it doesn't exist
+        if not is_redis_cache():
+            if cache.get(self.cache_key) is None:
+                cache.set(self.cache_key, self.attempts, self.window)
+        else:
+            cast("RedisCache", cache).set(self.cache_key, self.attempts, self.window, nx=True)
+    
+    def would_block(self) -> bool:
+        """Check if rate limit would be exceeded without decrementing counter."""
+        
+        current = cache.get(self.cache_key, 0)
+        return current <= 0 
+    
+    def decrement(self) -> None:
+        """Decrement rate limit counter for an email that will be sent."""
+        
+        with suppress(ValueError):
+            cache.decr(self.cache_key)
```

Comment:
This class duplicates code from the current rate limiter, there should be just implementation of that.

Reply:
Totally re-worked, no more duplication